sources/api/shibaken/src/warmpool/autoscaling_warm_pool.rs

```diff
@@ -0,0 +1,147 @@
+/// This module contains utilities for waiting until the instance on warm pool to be
+/// marked as InService.
+use crate::warmpool::error::{self, Result};
+use argh::FromArgs;
+use imdsclient::ImdsClient;
+use serde::Deserialize;
+use snafu::{OptionExt, ResultExt};
+use std::fs;
+use std::path::Path;
+use tokio::time::{sleep, Duration};
+
+//Retry strategy variables
+const FETCH_IMDS_SLEEP_DURATION: Duration = Duration::from_secs(5);
+const MAX_ATTEMPTS: i8 = 5;
+
+// Path to config file containing host's autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/warm-pool-wait.toml";
+const FETCH_LIFECYCLE_INTERVAL_SECS: Duration = Duration::from_secs(5);
+
+#[derive(FromArgs, Debug)]
+#[argh(subcommand, name = "warm-pool-wait")]
+/// Wait until the instance reaches the `InService` state
+pub(crate) struct WarmPoolWait {}
+
+impl WarmPoolWait {
+    pub(crate) async fn run(self) -> Result<()> {
+        let config_parse = get_config().await?;
+
+        let should_wait_value = config_parse.should_wait;
+        let marker_file_path = config_parse.marker_path;
+
+        println!("autoscaling.should-wait value is {}", should_wait_value);
+        if should_wait_value {
+            wait_until_inservice().await?;
+        }
+
+        fs::write(&marker_file_path, "").unwrap_or_else(|e| {
+            log::warn!("Failed to create marker file '{}', warm-pool-wait service may unexpectedly run again: '{}'",
+            &marker_file_path, e);
+        });
+        println!("Marker file path is {}", marker_file_path);
+
+        Ok(())
+    }
+}
+
+/// Uses an imdsclient function to fetch the lifecyclestate of the host
+async fn get_lifecycle_state(client: &mut ImdsClient) -> Result<String> {
+    let mut attempts: i8 = 0;
+
+    // Retry up to 5 times in case the lifecycle state is't ready (return 404); Waiting 5 seconds between each attempt.
+    while attempts < MAX_ATTEMPTS {
+        match client
+            .fetch_autoscaling_lifecycle_state()
+            .await
+            .context(error::ImdsRequestSnafu)?
+        {
+            Some(target_lifecycle_state) => return Ok(target_lifecycle_state),
+            None => {
+                sleep(FETCH_IMDS_SLEEP_DURATION).await;
+                attempts += 1;
+                continue;
+            }
```

Comment:
I'd prefer to have this function return a `Result<Option<String>>` and have the caller handle the waiting and retries - otherwise we're stacking up a lot of retry logic. There's already some retrying going on in `imdsclient`.

The caller can handle checking whether it's `None` (from a 404) or the desired string.